### PR TITLE
Arya game

### PR DESCRIPTION
**Pull request by aryanagarkar**

### CODE REVIEW

File: Story/AnnotationEnum.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace service
+{
+    public enum AnnotationEnum
+    {
+        success,
```
Comment (nagarkar) [inline Story/AnnotationEnum.cs:5]: make annotations have consistent case.

File: Story/GraphNode.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Service.Util;
+using System;
+
+namespace service
+{
+    public class GraphNode
+    {
+        public string character { get; set; }
```
Comment (nagarkar) [inline Story/GraphNode.cs:10]: case of property names should be consistent.

File: Story/SaveAndLoadStory.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                File.Delete(path);
+            }
+            String jsonString = JsonSerializer.Serialize(story);
+            File.WriteAllText(path, jsonString);
+        }
+
+        public static Graph loadStory(String path)
+        {
+            String jsonString = File.ReadAllText(path);
+            //Console.WriteLine(jsonString);
```
Comment (nagarkar) [inline Story/SaveAndLoadStory.cs:22]: Remove commented out code.

File: Story/Tests/MDictionaryTests.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    public class MDictionaryTests
+    {
+
+        [Test]
+        public void testMDictionariesEqual_SameOrder()
+        {
+            // Data setup
+                        
+            // Expectations
+            MDictionary<string, string> dict1 = new MDictionary<string, string>();
```
Comment (nagarkar) [inline Story/Tests/MDictionaryTests.cs:17]: Is this expectations or data setup?

File: Story/Tests/MDictionaryTests.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+            Assert.AreEqual(dict1, dict2);
+        }
+
+        [Test]
+        public void testMDictionariesEqual_differentOrder()
+        {
+            // Data setup
+                        
+            // Expectations
+            MDictionary<string, string> dict1 = new MDictionary<string, string>();
```
Comment (nagarkar) [inline Story/Tests/MDictionaryTests.cs:35]: ditto

File: Story/Tests/MDictionaryTests.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            MDictionary<string, string> dict2 = new MDictionary<string, string>();
+            dict2.Add("2", "node1");
+            dict2.Add("1", "start");
+
+            // Test and Assert
+            Assert.AreEqual(dict1, dict2);
+        }
+
+        [Test]
+        public void testMDictionariesNotEqual_differentNumberOfElements()
```
Comment (nagarkar) [inline Story/Tests/MDictionaryTests.cs:48]: Need more tests.
same number of elements, different values - should not be equal

**Review by nagarkar — CHANGES_REQUESTED** [review]

File: Story/Tests/SaveAndLoadStoryTests.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+        [Test]
+        public void testLoadedStoryGraphNodes()
+        {
+            // Data setup
+                        
+            // Expectations
+            Graph expected = new Graph();
+
+            GraphNode node1 = new GraphNode();
```
Comment (nagarkar) [inline Story/Tests/SaveAndLoadStoryTests.cs:20]: Create a helper function to encapsulate creation of a node.

File: Story/Util/MDictionary.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Service.Util
+{
+    public class MDictionary<TKey, TValue> : Dictionary<TKey, TValue>
+    {
+        public MDictionary() : base()
+        {
```
Comment (nagarkar) [inline Story/Util/MDictionary.cs:9]: Is this due to autoformatting? Does public Mdicionary(): base() {} not work?

File: Story/Util/MDictionary.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        public MDictionary() : base()
+        {
+
+        }
+
+        public override bool Equals(object obj)
+        {
+            MDictionary<TKey, TValue> other = obj as MDictionary<TKey, TValue>;
+            if (Count == other.Count)
```
Comment (nagarkar) [inline Story/Util/MDictionary.cs:16]: Instead, write
if (Count != other.Count) { return false; }
...

reduces indentation in the code and makes it more readable.

File: Story/Util/MDictionary.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                for (int i = 0; i < Count; i++)
+                {
+                    TKey key = Keys.ElementAt(i);
+                    if (!other.ContainsKey(key))
```
Comment (nagarkar) [inline Story/Util/MDictionary.cs:21]: Write in a more readable way, witha  helper:
if (!other.ContainsKey(key)  || !valuesEqualWithOtherMap(key, other)) {
    return false;
}

// valuesEqualWithOtherMap will encapsulate the code to compare the value for the key between the two maps.

**Review by nagarkar — CHANGES_REQUESTED** [review]

File: Story/GraphNode.cs (hunk 2 above)
Comment (aryanagarkar) [reply Story/GraphNode.cs:10]: Made all properties uppercase.

File: Story/SaveAndLoadStory.cs (hunk 3 above)
Comment (aryanagarkar) [reply Story/SaveAndLoadStory.cs:22]: Removed all commented out code and anything that I am not using.

File: Story/Tests/MDictionaryTests.cs (hunk 6 above)
Comment (aryanagarkar) [reply Story/Tests/MDictionaryTests.cs:48]: Added a test for the same number of elements and different values. All tests pass.

File: Story/Util/MDictionary.cs (hunk 8 above)
Comment (aryanagarkar) [reply Story/Util/MDictionary.cs:9]: It works, I changed it to public MDictionary() : base() { }.

File: Story/Util/MDictionary.cs (hunk 9 above)
Comment (aryanagarkar) [reply Story/Util/MDictionary.cs:16]: I changed this to if (Count != other.Count) { return false; }

File: Story/Util/MDictionary.cs (hunk 10 above)
Comment (aryanagarkar) [reply Story/Util/MDictionary.cs:21]: I wrote this helper method and made sure all the tests pass.

File: Story/Tests/MDictionaryTests.cs (hunk 4 above)
Comment (aryanagarkar) [reply Story/Tests/MDictionaryTests.cs:17]: I put the first one in Data setup and changed the names to make it clearer. The first MDictionary would be Data setup and the second would be expectations.

File: Story/Tests/MDictionaryTests.cs (hunk 5 above)
Comment (aryanagarkar) [reply Story/Tests/MDictionaryTests.cs:35]: I put the first one in Data setup and changed the names to make it clearer. The first MDictionary would be Data setup and the second would be expectations.

File: Story/Tests/SaveAndLoadStoryTests.cs (hunk 7 above)
Comment (aryanagarkar) [reply Story/Tests/SaveAndLoadStoryTests.cs:20]: I created a createNode() helper method and used it in my test. All tests pass.

File: Story/Tests/SaveAndLoadStoryTests.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+using Service.Util;
+
+namespace service.Tests
+{
+    [TestFixture]
+    public class SaveAndLoadStoryTests
+    {
+        string path = "../../../SampleStoryGraph.json";
+
+        private GraphNode createNode(CharacterTypeEnum character, List<string> conversations, MDictionary<string, List<string>> links){
```
Comment (nagarkar) [inline Story/Tests/SaveAndLoadStoryTests.cs:12]: Private methods after public methods.

File: Story/Tests/SaveAndLoadStoryTests.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        public void testLoadedStoryGraphNodes()
+        {
+            // Data setup
+                        
+            // Expectations
+            Graph expected = new Graph();
+
+            GraphNode startNode;
+            List<string> startNodeConversations = new List<string>();
+            startNodeConversations.Add("Hello");
```
Comment (nagarkar) [inline Story/Tests/SaveAndLoadStoryTests.cs:30]: creat elist inside helper.

File: Story/Tests/SaveAndLoadStoryTests.cs [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        private GraphNode createNode(CharacterTypeEnum character, List<string> conversations, MDictionary<string, List<string>> links){
+            GraphNode node = new GraphNode();
+            node.Links = links;
+            node.Character = character.ToString();
+            node.Conversations = conversations;
+            return node;
+        }
+
+        [Test]
+        public void testLoadedStoryGraphNodes()
```
Comment (nagarkar) [inline Story/Tests/SaveAndLoadStoryTests.cs:21]: Need another test where loaded story (actual) doesn't match with expected

File: Story/Util/MDictionary.cs [hunk 14]
```diff
@@ -0,0 +1,51 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace Service.Util
+{
+    public class MDictionary<TKey, TValue> : Dictionary<TKey, TValue>
+    {
+        public MDictionary() : base() { }
+
+        public bool valuesEqualWithOtherMap(TKey key, MDictionary<TKey, TValue> other)
```
Comment (nagarkar) [inline Story/Util/MDictionary.cs:10]: Should not be public. Move private methods to the end.

**Review by nagarkar — CHANGES_REQUESTED** [review]

File: Story/Util/MDictionary.cs (hunk 14 above)
Comment (aryanagarkar) [reply Story/Util/MDictionary.cs:10]: Changes to private and moved private method to the end.

File: Story/Tests/SaveAndLoadStoryTests.cs (hunk 11 above)
Comment (aryanagarkar) [reply Story/Tests/SaveAndLoadStoryTests.cs:12]: Went through all code, changed methods that are private to private, and moved all the private methods after public methods.

File: Story/Tests/SaveAndLoadStoryTests.cs (hunk 12 above)
Comment (aryanagarkar) [reply Story/Tests/SaveAndLoadStoryTests.cs:30]: I create an inside helper method that constructs the list and dictionary and passes it to the other helper method.

File: Story/Tests/SaveAndLoadStoryTests.cs (hunk 13 above)
Comment (aryanagarkar) [reply Story/Tests/SaveAndLoadStoryTests.cs:21]: I created another test where actual doesn't match with expected. All tests pass.

**Review by nagarkar — APPROVED** [review]

**Pull request merged**